File: python/xmlify/xml_to_json.py
"""Parse XML tool calls from model responses back to JSON with schema-aware type coercion."""

from __future__ import annotations

import re
from typing import Any

from .types import JsonSchema, ParsedToolCall, ToolDefinition


def parse_tool_calls(xml_text: str, tools: list[ToolDefinition]) -> list[ParsedToolCall]:
    """Parse <tool_call> blocks from the model's XML response text."""
    tool_map = {t.name: t for t in tools}
    calls: list[ParsedToolCall] = []

    for m in re.finditer(r'<tool_call\s+name="([^"]+)"[^>]*>([\s\S]*?)</tool_call>', xml_text):
        name = m.group(1)
        body = m.group(2)
        tool = tool_map.get(name)
        schema = tool.parameters if tool else None
        args = _parse_xml_element(body, schema)
        calls.append(ParsedToolCall(name=name, arguments=args))

    return calls


def _parse_xml_element(xml_body: str, schema: JsonSchema | None) -> dict[str, Any]:
    """Parse an XML element body into a dict, using schema for type coercion."""
    result: dict[str, Any] = {}

    # Self-closing tags
    for m in re.finditer(r"<([a-zA-Z_][a-zA-Z0-9_.-]*)(?:\s[^>]*)?\s*/>", xml_body):
        result[m.group(1)] = None

    # Collect tag names (skip self-closing)
    names: set[str] = set()
    for m in re.finditer(r"<([a-zA-Z_][a-zA-Z0-9_.-]*)(?:\s[^>]*)?>", xml_body):
        before = xml_body[: m.end()]
        if not before.endswith("/>"):
            names.add(m.group(1))

    for name in names:
        if name == "item":
            continue

        escaped = re.escape(name)
        content_match = re.search(
            rf"<{escaped}(?:\s[^>]*)?>([\s\S]*?)</{escaped}>", xml_body
        )
        if content_match:
            prop_schema = (schema or {}).get("properties", {}).get(name)
            result[name] = _parse_value(content_match.group(1), prop_schema)

    return result


def _parse_value(content: str, schema: JsonSchema | None) -> Any:
    """Parse a value from XML content, using schema for type coercion."""
    trimmed = content.strip()

    # Object type
    if schema and schema.get("type") == "object" and schema.get("properties"):
        return _parse_xml_element(trimmed, schema)

    # Array type
    if schema and schema.get("type") == "array":
        return _parse_array(trimmed, schema.get("items"))

    # No schema — try to detect structure
    if re.search(r"<[a-zA-Z_]", trimmed) and not (schema and schema.get("type")):
        if re.search(r"<item[\s>/]", trimmed):
            return _parse_array(trimmed, None)
        return _parse_xml_element(trimmed, schema)

    return _coerce_scalar(trimmed, schema)


def _parse_array(content: str, item_schema: JsonSchema | None) -> list[Any]:
    """Parse array values from XML content."""
    items: list[Any] = []

    for m in re.finditer(r"<item(?:\s[^>]*)?\s*/>|<item(?:\s[^>]*)?>([\s\S]*?)</item>", content):
        if m.group(1) is None:
            items.append(None)
        else:
            items.append(_parse_value(m.group(1), item_schema))

    # Fallback: comma-separated, JSON brackets, dimension strings
    if not items and content.strip():
        raw = content.strip()

        # Strip JSON-style brackets
        if raw.startswith("[") and raw.endswith("]"):
            raw = raw[1:-1]

        # Try comma split, then " x " (dimension pattern)
        parts = [s.strip() for s in raw.split(",") if s.strip()]
        if len(parts) <= 1 and re.search(r"\bx\b", raw, re.IGNORECASE):
            parts = [s.strip() for s in re.split(r"\s*x\s*", raw, flags=re.IGNORECASE) if s.strip()]

        for part in parts:
            # Strip JSON-style quotes
            cleaned = re.sub(r'^["\'](.*)["\']\s*$', r"\1", part)
            # Extract leading number if schema expects numeric
            if item_schema and item_schema.get("type") in ("number", "integer"):
                num_match = re.match(r"^-?\d+(?:\.\d+)?", cleaned)
                if num_match:
                    cleaned = num_match.group(0)
            items.append(_coerce_scalar(cleaned, item_schema))

    return items


def _coerce_scalar(value: str, schema: JsonSchema | None) -> Any:
    """Coerce a string value to the appropriate Python type."""
    unescaped = _unescape_xml(value)
    lower = unescaped.lower()

    schema_type = schema.get("type") if schema else None

    if not schema_type:
        # Best-effort inference
        if lower == "true":
            return True
        if lower == "false":
            return False
        if lower == "null" or unescaped == "":
            return None
        try:
            num = float(unescaped)
            if num == int(num) and "." not in unescaped:
                return int(num)
            return num
        except ValueError:
            return unescaped

    if schema_type in ("number", "integer"):
        try:
            n = float(unescaped)
            return int(n) if schema_type == "integer" else n
        except ValueError:
            return 0
    if schema_type == "boolean":
        return lower == "true" or unescaped == "1"
    if schema_type == "null":
        return None
    return unescaped


def _unescape_xml(s: str) -> str:
    return (
        s.replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", '"')
        .replace("&apos;", "'")
        .replace("&amp;", "&")
    )

File: src/xml-to-json.ts
import type { JsonSchema, ParsedToolCall, ToolDefinition } from "./types.js";

/**
 * Parse tool calls from the model's XML response text.
 *
 * Expects the model to emit tool calls like:
 * ```xml
 * <tool_call name="browser">
 *   <url>https://example.com</url>
 *   <action>click</action>
 * </tool_call>
 * ```
 *
 * Returns parsed JSON objects with typed values (using the schema
 * to coerce strings to numbers/booleans where appropriate).
 */
export function parseToolCalls(
  xmlText: string,
  tools: ToolDefinition[],
): ParsedToolCall[] {
  const toolMap = new Map(tools.map((t) => [t.name, t]));
  const calls: ParsedToolCall[] = [];

  // Match all <tool_call> blocks
  const callPattern =
    /<tool_call\s+name="([^"]+)"[^>]*>([\s\S]*?)<\/tool_call>/g;
  let match: RegExpExecArray | null;

  while ((match = callPattern.exec(xmlText)) !== null) {
    const name = match[1];
    const body = match[2];
    const tool = toolMap.get(name);
    const schema = tool?.parameters;
    const args = parseXmlElement(body, schema);
    calls.push({ name, arguments: args });
  }

  return calls;
}

/**
 * Parse a single XML element body into a JSON object,
 * using the schema to coerce types.
 */
function parseXmlElement(
  xmlBody: string,
  schema?: JsonSchema,
): Record<string, unknown> {
  const result: Record<string, unknown> = {};

  // Match self-closing tags: <name/> or <name attr/>
  const selfClosingPattern =
    /<([a-zA-Z_][a-zA-Z0-9_.-]*)(?:\s[^>]*)?\s*\/>/g;
  let match: RegExpExecArray | null;

  while ((match = selfClosingPattern.exec(xmlBody)) !== null) {
    result[match[1]] = null;
  }

  // Match content tags: <name>...</name>
  // Use a function to find matching open/close pairs at the same nesting level
  const tagStartPattern = /<([a-zA-Z_][a-zA-Z0-9_.-]*)(?:\s[^>]*)?>(?!.*\/>)/g;

  // Simpler approach: match <name>content</name> allowing nested same-name tags
  const names = new Set<string>();
  const namePattern = /<([a-zA-Z_][a-zA-Z0-9_.-]*)(?:\s[^>]*)?>/g;
  while ((match = namePattern.exec(xmlBody)) !== null) {
    // Skip if this is inside a self-closing tag
    const before = xmlBody.substring(0, match.index + match[0].length);
    if (!before.endsWith("/>")) {
      names.add(match[1]);
    }
  }

  for (const name of names) {
    // Find first <name...>content</name> for this element name
    const contentPattern = new RegExp(
      `<${escapeRegex(name)}(?:\\s[^>]*)?>([\\s\\S]*?)<\\/${escapeRegex(name)}>`,
    );
    const contentMatch = contentPattern.exec(xmlBody);
    if (contentMatch) {
      const propSchema = schema?.properties?.[name];
      result[name] = parseValue(contentMatch[1], propSchema);
    }
  }

  return result;
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/**
 * Parse a value from XML content, using schema for type coercion.
 */
function parseValue(content: string, schema?: JsonSchema): unknown {
  const trimmed = content.trim();

  // Object type — recurse
  if (schema?.type === "object" && schema.properties) {
    return parseXmlElement(trimmed, schema);
  }

  // Array type — collect <item> children
  if (schema?.type === "array") {
    return parseArray(trimmed, schema.items);
  }

  // If content contains child elements and no explicit schema, try object parse
  if (/<[a-zA-Z_]/.test(trimmed) && !schema?.type) {
    // Could be array of <item> or nested object
    if (/<item[\s>\/]/.test(trimmed)) {
      return parseArray(trimmed, undefined);
    }
    return parseXmlElement(trimmed, schema);
  }

  // Scalar — coerce based on schema type
  return coerceScalar(trimmed, schema);
}

function parseArray(content: string, itemSchema?: JsonSchema): unknown[] {
  const items: unknown[] = [];

  // Match <item> elements
  const itemPattern = /<item(?:\s[^>]*)?\s*\/>|<item(?:\s[^>]*)?>([\s\S]*?)<\/item>/g;
  let match: RegExpExecArray | null;

  while ((match = itemPattern.exec(content)) !== null) {
    if (match[1] === undefined) {
      items.push(null); // self-closing <item/>
    } else {
      items.push(parseValue(match[1], itemSchema));
    }
  }

  return items;
}

function coerceScalar(value: string, schema?: JsonSchema): unknown {
  const unescaped = unescapeXml(value);

  if (!schema?.type) {
    // Best-effort type inference
    if (unescaped === "true") return true;
    if (unescaped === "false") return false;
    if (unescaped === "null" || unescaped === "") return null;
    const num = Number(unescaped);
    if (!isNaN(num) && unescaped !== "") return num;
    return unescaped;
  }

  switch (schema.type) {
    case "number":
    case "integer":
      return Number(unescaped);
    case "boolean":
      return unescaped === "true" || unescaped === "1";
    case "null":
      return null;
    default:
      return unescaped;
  }
}

function unescapeXml(s: string): string {
  return s
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&amp;/g, "&");
}
